iterator.go

```diff
@@ -0,0 +1,162 @@
+// Licensed to The Moov Authors under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. The Moov Authors licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package ach
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+	"unicode"
+)
+
+// Iterator is a data structure for processing an ACH file one entry at a time.
+type Iterator struct {
+	reader     *Reader
+	scanner    *bufio.Scanner
+	cachedLine string
+}
+
+// NewIterator returns an Iterator
+func NewIterator(r io.Reader) *Iterator {
+	reader := NewReader(strings.NewReader("")) // the input is not used, we rely on .readLine()
+	reader.skipBatchAccumulation = true        // don't call .AddBatch(..)
+
+	out := &Iterator{
+		reader:  reader,
+		scanner: bufio.NewScanner(r),
+	}
+	return out
+}
+
+func (i *Iterator) SetValidation(opts *ValidateOpts) {
+	if i.reader != nil {
+		i.reader.SetValidation(opts)
+	}
+}
+
+// GetHeader will return the FileHeader once encountered by the iterator.
+// Call NextEntry() at least once to populate the header.
+func (i *Iterator) GetHeader() *FileHeader {
+	if i.reader != nil {
+		return &i.reader.File.Header
+	}
+	return nil
+}
+
+// GetControl will return the FileControl once encountered by the iterator.
+// Call NextEntry() at least once to populate the control.
+func (i *Iterator) GetControl() *FileControl {
+	if i.reader != nil {
+		return &i.reader.File.Control
+	}
+	return nil
+}
+
+// NextEntry will return the next available EntryDetail record and the BatchHeader the entry belongs to.
+//
+// IAT entries are not currently supported.
+func (i *Iterator) NextEntry() (*BatchHeader, *EntryDetail, error) {
+	// Clear the reader's File
+	defer func() {
+		i.reader.File = File{}
+	}()
+
+	// Read the file one line at a time
+	line := i.cachedLine
+	if line != "" {
+		i.cachedLine = "" // clear cache
```

Comment:
Why wouldn't a pre-existing cachedLine be an error state?

Reply:
EntryDetail records can have addendas which we want to include on the EntryDetail. Since a `bufio.Scanner` has no method to peek at the next `.Text()` result we have to cache the line if it's not an addenda. On the next iteration we want to take a cached line (and skip calling `.Text()` for that iteration) when it exists.